RFC:  PCIE: MSI: Fixes

finikorg opened the pull request:
Various MSI fixes

finikorg committed:
pcie: Treat index parameter as dword

Make index parameter as dword index. This only makes difference for
accessing BAR index after 64 bit BAR registers.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
pcie: shell: Print 64 bit BARs

Print full 64 bit BARs.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
pcie: shell: Print more MSI information

For pcie ls command print more detailed MSI / MSI-X information.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

jhedberg commented on drivers/pcie/host/shell.c:
```diff
@@ -45,15 +72,21 @@ static void show_bars(const struct shell *shell, pcie_bdf_t bdf)
-			      PCIE_CONF_BAR_ADDR(data));
+			      PCIE_CONF_BAR_64(data) ? ", 64-bit" : "");
+
+		shell_fprintf(shell, SHELL_NORMAL, " addr ");
```
maybe `" addr 0x"`?

dcpleung commented on drivers/pcie/host/shell.c:
```diff
@@ -29,7 +29,34 @@ static void show_msi(const struct shell *shell, pcie_bdf_t bdf)
+
+		shell_fprintf(shell, SHELL_NORMAL,
+			      "    MSI-X support%s table size %d\n",
+			      (data & PCIE_MSIX_MCR_EN) ? ", enabled" : "",
```
", disabled" when false?

finikorg commented on drivers/pcie/host/shell.c:
```diff
@@ -45,15 +45,21 @@ static void show_bars(const struct shell *shell, pcie_bdf_t bdf)
+				      pcie_conf_read(bdf, bar));
 		}
+
+		shell_fprintf(shell, SHELL_NORMAL, "%08x\n",
+			      PCIE_CONF_BAR_ADDR(data));
```
08 means number of characters, not bytes.

finikorg committed:
pcie: shell: Add disable indicator

Add disable indicator if MSI or MSI-X are disabled.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
pcie: shell: Add leading 0x for hexadecimal address

Add leading 0x to address prints.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg commented on drivers/pcie/host/pcie.c:
```diff
@@ -83,6 +83,7 @@ bool pcie_get_mbar(pcie_bdf_t bdf, unsigned int index, struct pcie_mbar *mbar)
 
 		if (PCIE_CONF_BAR_MEM(addr) && PCIE_CONF_BAR_64(addr)) {
 			reg++;
+			index--; /* treat index as dword */
```
This change we still need to verify

jhedberg replied:
How and when will you verify it?

finikorg replied:
Added patch with documentation for the index parameter.

finikorg replied:
@tbursztyka Could you review this change.

tbursztyka replied:
This is not right, whether we are with 32 bits BARs or 64 bits ones: index is the same. Thus why we have this variable reg for actually moving through.

index is between 1 and 6 (to target bar 0 to bar 5, not ideal but...). Where reg is between 4 to 15 in 64 bits (4 to 9 in 32).

finikorg replied:
We have a problem with it with MSI-X, when we need to access register with the index in dwords. So what is the current usage? You just take index 0? It probably does not make a difference.

finikorg replied:
@tbursztyka the idea is to fix the issue #37444.

finikorg commented:
@najumon1980

finikorg committed:
pcie: Document pcie_get_mbar() index parameter

Document that index is measured in DWORD and 64 bit register consumes
2 indices.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

dcpleung commented on drivers/pcie/host/shell.c:
```diff
@@ -29,7 +29,34 @@ static void show_msi(const struct shell *shell, pcie_bdf_t bdf)
+			      PCIE_MSIX_MCR_TSIZE_SHIFT) + 1;
+
+		shell_fprintf(shell, SHELL_NORMAL,
+			      "    MSI-X support%s table size %d\n",
```
I think this should be `MSI-X support %s, table size %d\n` instead.

Also, should it be showing table size if MSI-X is disabled?

finikorg replied:
Why not? it is the same in lspci

jhedberg replied:
Because it looks less strange/more readable. I.e. please fix it in lspci as well (separate patch or PR though)

finikorg replied:
I mean lspci in Linux. I think it does make sense to print this information even if it is disabled.

jhedberg replied:
> I mean lspci in Linux. I think it does make sense to print this information even if it is disabled.

I dont think anyone requested not to print something. My comment was just about how to split up the format string.

tbursztyka commented on drivers/pcie/host/msi.c:
```diff
@@ -278,8 +277,11 @@ bool pcie_msi_enable(pcie_bdf_t bdf,
 			disable_msi(bdf, base);
 		}
 
+		if (base_msix != 0U) {
```
refactor this with previous if block as well, so:
```
if (base_msix != 0) {
        if (base != 0) {
                disable_msi();
        }

        base = base_msix;
        msi = false;
}
```

finikorg replied:
fixed this way

tbursztyka commented on drivers/pcie/host/shell.c:
```diff
@@ -29,7 +29,34 @@ static void show_msi(const struct shell *shell, pcie_bdf_t bdf)
 
 	if (msi) {
-		shell_fprintf(shell, SHELL_NORMAL, "    MSI-X support\n");
+		uint32_t offset, table_size;
```
your commit message talks about MSI but it's MSI-X here.

tbursztyka commented on include/drivers/pcie/pcie.h:
```diff
@@ -99,6 +99,7 @@ extern bool pcie_probe(pcie_bdf_t bdf, pcie_id_t id);
  * Value of n has to be below 6, as there is a maximum of 6 BARs. The indices
  * are order-preserving with respect to the endpoint BARs: e.g., index 0
  * will return the lowest-numbered memory BAR on the endpoint.
+ * Index is measured in DWORD: e.g., 64 bit register consumes 2 indices.
```
no need of that patch, see my comment in the previous patch you did on index.

finikorg committed:
msi: Fix overwriting base register in msi_allocate()

When CONFIG_PCIE_MSI_X option is enabled we always overwrite
base register value.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
msi: Fix overwriting base register in msi_enable()

When CONFIG_PCIE_MSI_X is enabled, we should set base to msix only
if we find that it is supported.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
Merge branch 'msi-fix' into msi

tbursztyka commented on drivers/pcie/host/pcie.c:
```diff
@@ -83,6 +83,7 @@ bool pcie_get_mbar(pcie_bdf_t bdf, unsigned int index, struct pcie_mbar *mbar)
 
```
Ok I see why pcie_get_mbar() is not doing what is required for MSI-X BIR (which what triggered this patch to be created in the first place).

index != bar index. And that comes back from the first commit adding this function (commit id e0390535467)
Index is not a BAR index but a valide BA index.
And it's documented that way "The indices are order-preserving with respect to the endpoint BARs: e.g., index 0 will return the lowest-numbered memory BAR on the endpoint."

Let's take an example of 3 memory Base Address. Here in 32 bits only:

- BA 0 (BAR index 0)
- BA 1 (BAR index 1)
- BA 2 (BAR index 2)

Let's say now that BA 1 is in 64 bits, it will be:

- BA 0 (BAR index 0)
- BA 1 (BAR index 1)
- BA 2 (BAR index 3 <- because BAR index 2 is also part of BA 1 it's higher address part)

So that's what index represents here. BA x, not BAR index x.
And it's not practical when it comes to MSI-x with it's BIR: "Base address register Indicator Register". Indicator can be interpreted as Index as well.

So pcie_get_mbar() should be kept as is. The proposed patch is not valid (it break what pcie_get_mbar() is supposed to do, according to its documentation). 

I'll came a fix for msi-x BIR.

tbursztyka commented on drivers/pcie/host/shell.c:
```diff
@@ -79,7 +79,7 @@ static void show_bars(const struct shell *shell, pcie_bdf_t bdf)
 			      PCIE_CONF_BAR_64(data) ? ", 64-bit" : "");
 
-		shell_fprintf(shell, SHELL_NORMAL, " addr ");
+		shell_fprintf(shell, SHELL_NORMAL, " addr 0x");
```
this patch is correcting a previous patch in the same PR, so merge it directly with the original one.

finikorg commented:
Cherry-picked relevant patches to other PRs.
